source/common/http/http2/codec_impl.cc

```diff
@@ -577,12 +587,10 @@ int ConnectionImpl::onFrameReceived(const nghttp2_frame* frame) {
     }
   }
 
-  // Only raise GOAWAY once, since we don't currently expose stream information. Shutdown
-  // notifications are the same as a normal GOAWAY.
-  if (frame->hd.type == NGHTTP2_GOAWAY && !raised_goaway_) {
+  // Shutdown notifications are the same as a normal GOAWAY.
+  if (frame->hd.type == NGHTTP2_GOAWAY) {
```

Comment:
I'm not sure if this is fine or if we want to track which goaway error codes we've already seen, and run the callbacks once per error code.

Reply:
It is possible to receive GOAWAY frames with different error codes:

> An endpoint MAY send multiple GOAWAY frames if circumstances change. For instance, an endpoint that sends GOAWAY with NO_ERROR during graceful shutdown could subsequently encounter a condition that requires immediate termination of the connection.

(from https://http2.github.io/http2-spec/#GOAWAY)

Reply:
Hmm in that case should the logic for failing the HC be that we see a NO_ERROR + no error GOAWAYs?

Also I'm not sure what the implications are of making `onGoAway` trigger multiple times per stream, maybe @mattklein123 or @alyssawilk has an idea? I'd be worried about invariants elsewhere relying on this being called at most once.

Reply:
> Hmm in that case should the logic for failing the HC be that we see a NO_ERROR + no error GOAWAYs?

Hm, I don't quite follow, could you rephrase?

> I'd be worried about invariants elsewhere relying on this being called at most once.

Agreed. It seems it was added in https://github.com/envoyproxy/envoy/pull/103. I'm not sure if that was an optimization back then. I'm also not sure if it has become an expected behaviour (intentionally or not) since then.

Reply:
```
  if (request_encoder_ && error_code == Http::GoAwayErrorCode::NoError) {
    received_no_error_goaway_ = true;
    return;
  }
```

if a GOAWAY error follows a NO_ERROR then `received_no_error_goaway_` would still be true. I was thinking that the logic should reset this flag if it saw an error frame.

Reply:
Ah, I see. I don't expect that to change the behaviour as it stands:
* If there isn't an active probe, the first (NO_ERROR) GOAWAY would've caused the connection to be closed (and we only set `received_no_error_goaway_ = true;` when there's an active probe). 
* If there is an active probe, the stream reset will end up resetting state (`resetState()` will be called by `onResetStream()`).
* In general, I believe a connection close will end up resetting state because the `onResetStream()` callbacks will run ([from here](https://github.com/envoyproxy/envoy/blob/0125d6fd611958fdef85e6665f307d2bffd2a531/source/common/http/codec_client.cc#L71)).

I admit this isn't obvious / does feel a bit brittle -- happy to add a comment and/or reset the flag anyways.

Reply:
Friendly ping @mattklein123 @alyssawilk

Reply:
Sorry for the delay. My preference would be to not raise multiple goaways from the codec as I'm guessing there will be code that won't expect this and will break.

Can we just keep the previous behavior and not handle the case where multiple go away frames are received? We can potentially deal with this in a follow up if someone requests it?

Reply:
Thanks for the feedback - that seems reasonable to me. I've reverted that specific commit and added a TODO instead.